---
title: "End Joining Signatures - dev2"
author: "Charlie Soeder"
date: "3/25/2019"
output: pdf_document
bibliography: references.bib

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)


library("readr")
library("yaml")
library("knitr")
library("tidyverse")


```

##25 March 2019
From Talia:

>If Danny's progeny are from two types of parental crosses, I guess I would just do analysis on one of those crosses, just to keep it easier.  Find Indels in those progeny, then compare to mcm5a7 progeny.

Ok, going to restrict to w3 and w4.

Maybe redo the samples summary to reflect this?


>Males were numbered based on whether their father was homozygous w1118 or Canton-S and the number of their het- erozygous mother. For example, male cs12.3 had a Canton-S father, its mother was female number 12, and it was the third male selected for DNA extraction. Sibling numbers may not be continuous, as males with low DNA concentrations after DNA extraction were not selected for sequencing.


```{r include=FALSE}

trammel <- read_yaml("../config.yaml")
data_sets.df <- plyr::ldply(trammel$data_sets, data.frame)

```


```{r echo=FALSE}


data_sets.df.sparse <- data_sets.df %>% filter(subgroups=='all') %>% select(-c(subgroups,sra)) %>% distinct() 

data_sets.df.sparse.crossMeta <- data_sets.df.sparse %>% filter(experimental == 'control') %>% select(c(name, source, pedigree, sex))  %>% mutate(tmp=name) %>% separate(tmp, sep="_", into = c("mating","offspring_id"))  %>% mutate(tmp=mating) %>% separate(tmp, sep="[sw]", into=c("A","B"))%>% mutate(tmp=mating) %>% separate(tmp, sep="[0-9]", into=c("A","C")) %>% mutate(mating=case_when(pedigree=='child' ~ mating), father_type=case_when(pedigree=='child' ~ A), mother_id=case_when(pedigree=='child' ~ B) ) %>% select(-c(A,B,C)) %>% mutate(father_type=case_when(father_type=='w' ~ "w1118",father_type=='cs' ~ "CantonS") ) 


```

```{r echo=FALSE}


data_sets.df.sparse.crossMeta %>% arrange(pedigree, desc(name)) %>% kable(caption="Sequenced Samples from Control Cross")
```
```{r echo=FALSE}

data_sets.df.sparse.crossMeta %>% filter(pedigree=='child') %>% group_by(father_type) %>% summarize(count=n()) %>% kable(caption="Control Cross Samples by Cross Type")

```

```{r echo=FALSE}

data_sets.df.sparse.crossMeta %>% filter(pedigree=='child') %>% group_by(father_type, mother_id) %>% summarize(count=n()) %>% kable(caption="Number of Male Offsping Sequenced, by Cross Type and Female ID")

```

It might be good to go ahead and do a YAML for all the flies, with the ones being ignored in a null subgroup. Write a rule to download based on provided SRAs. 



## 28 March 2019

Ok cool I finally got some slots on Longleaf and the VCFs are building. using temporaries right now (variants called to date).

It occurs to me I may have to go back and rework some reporting rules to reflect the comparison of two different subrgoup variants..........


```{bash eval=FALSE}
$ pwd
/Users/csoeder/Research/EJgrepper/dev/meta/VCFs
$ prefix=control
$ cat control.vs_dm6.bwaUniq.summary | sed -e 's/^/'$prefix'\t/g'> ../all_groups.vs_dm6.bwaUniq.calledVariants.summary
$ prefix=mutant
$ cat mutant.vs_dm6.bwaUniq.summary | sed -e 's/^/'$prefix'\t/g' >> ../all_groups.vs_dm6.bwaUniq.calledVariants.summary
```




```{r include=FALSE}
refGenomes_summary_df <- read_delim("meta/reference_genomes.summary", 
    "\t", escape_double = FALSE, col_names = FALSE, 
    trim_ws = TRUE)

names(refGenomes_summary_df) <- c("refGenome","measure","value")

```
```{r echo=FALSE, include=FALSE}
human_readable_croncher <- function(num_in) {
	dig <- 3
	num_out <- formatC(num_in, digits=dig, format='g') %>% as.numeric() %>% sitools::f2si()
	return(num_out)
}

allGroups.vs_dm6.calledVariants.summary <- read_delim("meta/all_groups.vs_dm6.bwaUniq.calledVariants.summary", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(allGroups.vs_dm6.calledVariants.summary) <- c("group", "aligner", "variant_type", "count")
allGroups.vs_dm6.calledVariants.summary$aligner <- as.factor(allGroups.vs_dm6.calledVariants.summary$aligner)
allGroups.vs_dm6.calledVariants.summary$variant_type <- as.factor(allGroups.vs_dm6.calledVariants.summary$variant_type)

reference_genomes_summaryStats.sprud <- refGenomes_summary_df %>%  spread(measure,value) %>%  select(c(-number_contigs)) 

allGroups.vs_dm6.calledVariants.summary.sprud <- allGroups.vs_dm6.calledVariants.summary %>% spread(variant_type, count) %>%  transmute( aligner=aligner, group = group, INDELs = total_indel_count, SNPs = total_snp_count) 

allGroups.vs_dm6.calledVariants.summary.sprud$total_bp <- reference_genomes_summaryStats.sprud$number_bases


allGroups.vs_dm6.calledVariants.summary.sprud %>%  transmute( aligner=aligner, INDELs = INDELs, INDEL_per_kb=1000*INDELs/total_bp, SNPs = SNPs, SNP_per_kb=1000*SNPs/total_bp) %>% mutate(INDELs=human_readable_croncher(SNPs), INDELs=human_readable_croncher(SNPs) ) %>%  kable(caption="Variant Counts and Frequency by Type and Group", digits=1 )
```



```{r echo=FALSE, include=FALSE}
mutant.calledVariants.bwaUniq.lmiss <- read_delim("meta/VCFs/mutant.vs_dm6.bwaUniq.summary.lmiss", "\t", escape_double = FALSE, trim_ws = TRUE)
mutant.calledVariants.bwaUniq.lmiss$group <- "mutant"

control.calledVariants.bwaUniq.lmiss <- read_delim("meta/VCFs/control.vs_dm6.bwaUniq.summary.lmiss", "\t", escape_double = FALSE, trim_ws = TRUE)
control.calledVariants.bwaUniq.lmiss$group <- "control"

allGroups.calledVariants.lmiss <- rbind(control.calledVariants.bwaUniq.lmiss, mutant.calledVariants.bwaUniq.lmiss) %>% select(c(group, N_MISS, N_DATA)) %>%  mutate(group=as.factor(group), N_PRES=floor(N_DATA/2))

```

This is a little different than before; uses the callable chromosome number instead of the sample count (since the two VCFs could conceivably have different sample numbers) and rounds down. 


```{r echo=FALSE}
library("ggbio")


nsamps <- max(allGroups.calledVariants.lmiss$N_PRES)

ggplot(allGroups.calledVariants.lmiss) + geom_freqpoly(aes(x=N_PRES, group=group, color=group), bins=nsamps) + scale_x_continuous(name ="Number Samples",limits=c(1,nsamps), breaks =seq(1,nsamps,1)) + theme_clear() + labs(title="Histogram of Variant Site Count,\n by Number of Samples Called At Site", y="Number of Sites")
```

## 29 March 2019

Ok cool VCFs built and summarized. 


Ugh, geom_text + filter apparently fails with a " Aesthetics must either be length one, or the same length as the data" when the filter results are empty

```{r eval=FALSE}

mutant.calledVariants.imiss <- read_delim("meta/VCFs/mutant.vs_dm6.bwaUniq.summary.imiss", "\t", escape_double = FALSE, trim_ws = TRUE)
mutant.calledVariants.imiss$group <- "mutant"

control.calledVariants.imiss <- read_delim("meta/VCFs/control.vs_dm6.bwaUniq.summary.imiss", "\t", escape_double = FALSE, trim_ws = TRUE)
control.calledVariants.imiss$group <- "control"


allGroups.calledVariants.imiss <- rbind(control.calledVariants.imiss, mutant.calledVariants.imiss)  %>%  mutate(name=as.factor(INDV), group=as.factor(group), N_PRES=N_DATA-N_MISS) %>% select(c( name, N_MISS, N_PRES, F_MISS,group))


```


```{r eval=FALSE}


allGroups.imiss.augmented <- inner_join(allGroups.calledVariants.imiss, all_alignments %>%  filter(measure=='total_breadth') %>% filter(aligner=="bwaUniq") %>% select(c(sample,value)) %>% rename(breadth=value) %>% mutate(breadth = 100*breadth), by=c("name"="sample"))

allGroups.imiss.augmented <-inner_join(allGroups.imiss.augmented, all_alignments %>%  filter(measure=='avg_depth') %>% filter(aligner=="bwaUniq") %>% select(c(sample,value)) %>% rename(depth=value), by=c("name"="sample"))

allGroups.imiss.augmented <- allGroups.imiss.augmented %>%  gather(breadth:depth, key="measure", value="value")

```


```{r eval=FALSE} 
ggplot(allGroups.imiss.augmented) + geom_point(aes(x= value, y=1-F_MISS, color=group, shape=group)) + facet_grid(.~measure, scales="free_x") + theme_clear() + labs(x="", y="Fraction of Sites Callable", title="Jointly Called Variants Callable per Sample,\n by Breadth and Depth of Coverage")

```


So with the reporting out of the way, onto analysis of the variants....... 

Rewriting the Winnower rule to preserve the filtered VCF, then run allele count, such that the VCF can then be used for a Novelist rule to remove sites that are variable in the parents. 

oh cool adventures in purr and broom! use map_df to specify a data frame output so that the columns of the glanced stat test can be manipluated dplyr-style:


```{r eval=FALSE}
insert_truth.Tbl <- filteredTbl.biallele %>% mutate(ins=delta_bp>0) %>% group_by(ins,group) %>% summarise(count=n()) %>% spread(ins,count)%>% rename(ins=`TRUE`, del=`FALSE`)

insert_truth.Tbl <- cbind(insert_truth.Tbl, map2(insert_truth.Tbl$del, insert_truth.Tbl$del+insert_truth.Tbl$ins, prop.test) %>% map_df(glance) %>% select(c("estimate","p.value")) %>% rename(fraction_del=estimate) )
```

## 1 Apr 2019

```{r eval=FALSE}
ggplot(filteredTbl.biallele) + geom_freqpoly(aes(x=mac, color=group),bins=31) + facet_wrap(chrom~.) + theme_clear() + labs(x="Minor Allele Count (chromosomes)", y="Number of Sites", title="Number of Quality INDELs, by Minor Allele Count and Chromosome")

```


Adding the qualification to "Novel Singeltons" that parent-derived alleles must get scrubbed.

Currently adding Site Ids by VCF surgery; maybe add them earlier??

Current approach: subset VCF to parent, select sites with parent having >0 alt alleles, kick those out.
Q1: what if this rule needs to be expanded to multiallelic sites? 
```{bash eval=FALSE}
grep "1|1\|0|1"
```
vs
```{bash eval=FALSE}
grep "[1-9]|[1-9]\|0|[1-9]"
```
?

Q2: what if the parent allele is A and the mutant allele is R? (weird back-mutation special case)
These are r

uh, the vcftools --snp/snps option doesn't seem to be working? I guess just use grep?



```{r echo=FALSE}

multicolumn_ACloader <- function(filename){
	
	max_fields<- count.fields(filename, sep = '\t') %>% max()

	tubb <- read.table(filename, header = FALSE, sep = "\t", fill = TRUE,col.names = paste0("V",seq_len(max_fields)))

	tubb.df <- tubb[1:7]

	names(tubb.df) <- c("site_id", "chrom", "pos", "var_count", "chroms_avail", "allele", "allele_count")
	tubb.df$var_num <- 1
	
	for (i in seq(8,max_fields,2)){
		tubb.df.tmp <- cbind(tubb[1:5], tubb[c(i,i+1)])
		names(tubb.df.tmp) <- c("site_id", "chrom", "pos", "var_count", "chroms_avail", "allele", "allele_count")
		tubb.df.tmp$var_num <- i/2-2
		tubb.df <- rbind(tubb.df, tubb.df.tmp %>% filter(!is.na(allele_count))  )
	}

	tubb.df <- tubb.df %>% mutate(allele_freq = allele_count/chroms_avail) %>% as_tibble()

	
	
		return( tubb.df)
}

control.novel.counts <- multicolumn_ACloader("analysis/control.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.novel.count")
control.novel.counts$group <- as.factor("control")


mutant.novel.counts <- multicolumn_ACloader("analysis/mutant.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.novel.count")
mutant.novel.counts$group <- as.factor("mutant")


biallele.novel.counts <- rbind(control.novel.counts, mutant.novel.counts) %>% gather(variable, value, -(site_id:chroms_avail), -var_num) %>% unite(temp, var_num, variable) %>%  spread(temp, value) %>% mutate(len1 = nchar(`1_allele`),len2 = nchar(`2_allele`), `1_allele_count`=as.numeric(`1_allele_count`), `2_allele_count`=as.numeric(`2_allele_count`), `1_allele_freq`=as.numeric(`1_allele_freq`), `2_allele_freq`=as.numeric(`2_allele_freq`)) 


biallele.novel.counts.tmp <- biallele.novel.counts %>% filter(`1_allele_count` >= `2_allele_count`) %>% mutate(delta_bp=len2-len1,  mac=`2_allele_count`)

biallele.novel.counts <- rbind(biallele.novel.counts.tmp, biallele.novel.counts %>% filter(`1_allele_count` < `2_allele_count`) %>% mutate(delta_bp=len1-len2, mac=`1_allele_count`)) %>% mutate( group=`1_group`) %>%  select(-c(`2_group`))

```



```{r echo=FALSE}
ggplot(biallele.novel.counts %>%  group_by(mac, delta_bp,group) %>% summarise(count=n())) + geom_tile(aes(x=mac,y=delta_bp, fill=count)) +facet_grid(group~.) + theme_clear() #+ labs(x="Minor Allele Count (chromosomes)", y="Nucleotides Inserted or Deleted", title="Heatmap of INDEL Count, by Frequency and Size") + scale_fill_gradient(name = "Site Count", trans = "log", breaks = c(1,10,100,1000), labels = c(1,10,100,1000)) + theme( panel.background = element_rect(fill = "black", colour = "black"))
```




```{r echo=FALSE}
ggplot(biallele.novel.counts) + geom_freqpoly(aes(x=mac),bins=15, color='black') + geom_freqpoly(aes(x=mac, color=chrom),bins=31) + facet_grid(group~.) + theme_clear() #+ labs(x="Minor Allele Count (chromosomes)", y="Number of Sites", title="Number of Quality INDELs, by Minor Allele Count, Chromosome, and Treatment")

```
```{r eval=FALSE}

biallele.novel.byChrom <- inner_join(biallele.novel.counts %>% group_by(group, chrom) %>% summarise(novel=n()), filteredTbl.biallele %>% group_by(group, chrom) %>% summarize(total=n()), by=c("group"="group", "chrom"="chrom") ) 

biallele.novel.byChrom %>%  mutate(percent=human_readable_croncher( 100*novel/total)) %>%  kable(caption="Hereditarily Novel Biallaeleic INDELs, by Treatment and Chromosome")
```





```{r echo=FALSE}
ggplot(biallele.novel.counts)  + geom_freqpoly(aes(x=delta_bp, color=group), bins=13)  + theme_clear() + labs(x="Nucleotides Inserted/Deleted", y="Number of Sites", title="INDEL Count by Size")
```



```{r eval=FALSE}
novel.insert_truth.Tbl <- biallele.novel.counts %>% mutate(ins=delta_bp>0) %>% group_by(ins,group) %>% summarise(count=n()) %>% spread(ins,count)%>% rename(ins=`TRUE`, del=`FALSE`)
novel.insert_truth.Tbl <- cbind(novel.insert_truth.Tbl, map2(novel.insert_truth.Tbl$del, novel.insert_truth.Tbl$del+novel.insert_truth.Tbl$ins, prop.test) %>% map_df(glance) %>% select(c("estimate","p.value")) %>% rename(fraction_del=estimate) )

kable(novel.insert_truth.Tbl, caption = "Deletion bias in mcm5 Mutants  (Hereditarily Novel)")

```




```{r echo=FALSE}

biallele.novel.counts.singleton <- biallele.novel.counts %>% filter(mac==1) 

biallele.novel.counts.singleton %>% group_by(group, chrom) %>% summarise(novel_singleton=n())
```

```{r eval=FALSE}

full_join(biallele.novel.byChrom, biallele.novel.counts.singleton %>% group_by(group, chrom) %>% summarise(novel_singleton=n()), by=c("group"="group", "chrom"="chrom")) %>% mutate(novel_singleton=case_when(is.na(novel_singleton) ~0, TRUE ~ as.double( novel_singleton)), percent_novel_singleton =100*novel_singleton/novel)  %>%  kable(caption="Hereditarily Novel, Singleton Biallaeleic INDELs, by Treatment and Chromosome")

```

Fortifying the VCF_Novelist rule

```{bash eval=FALSE}


vcftools --vcf {input.vcf_in}.anc.tmp {par_string} --recode --recode-INFO-all --stdout | grep -v "#" | grep -v "1|1\|0|1\|1|0" | cut -f 3 > {input.vcf_in}.hom0.list;



vcftools --remove-indv DfMcm5 --remove-indv Mcm5-A7 --vcf variants/mutant.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.vcf.anc.tmp  --recode --recode-INFO-all  --stdout | grep -v "#" | less



```

## 2 April 2019

Hmm, looks  like adding the rare ref alleles has dramatically changed the results: Many more novel variants and novel-singleton variants in the control now. (?!)

One possibility: there is residual difference between w3, w4, and w1118. Checking this my comparing the frequencies at novel sites: w3 and w4 would be expected to have little overlap if novel sites are due to 3 vs 4 differences

```{bash eval=FALSE}

vcftools --vcf control.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.novel.vcf --indv w3_25 --indv w3_1 --indv w3_12 --indv w3_8 --indv w3_16 --indv w3_14 --indv w3_6 --indv w3_15 --indv w3_24 --indv w3_13 --indv w3_26 --indv w3_5 --indv w3_9 --indv w3_21 --indv w3_4 --indv w3_17 --indv w3_18 --indv w3_11 --freq --stdout | tail -n +2 | tr ":" "\t" | cut -f 1,8 > control.w3.frq


vcftools --vcf control.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.novel.vcf --indv w4_3 --indv w4_16 --indv w4_4 --indv w4_15 --indv w4_11 --indv w4_17 --indv w4_13 --indv w4_1 --indv w4_2 --indv w4_12 --freq --stdout | tail -n +2 | tr ":" "\t" | cut -f 1,8 > control.w4.frq

paste control.w[34].frq | cut -f 1,2,4 >control.diff.frq

```

```{r echo=FALSE}

control_diff <- read_delim("control.diff.frq", "\t", escape_double = FALSE, col_names = FALSE, trim_ws = TRUE)
names(control_diff) <- c("chrom","w3","w4")
control_diff$chrom <- as.factor(control_diff$chrom)
control_diff$diff <- control_diff$w3 - control_diff$w4

ggplot(control_diff) + geom_point(aes(x=w3, y=w4, color=chrom))
ggplot(control_diff) + geom_violin(aes(x=chrom, y=diff, color=chrom))
```

Hmm, no looks like those AFs are pretty well correllated. 

## 3 April 2019

Okay to compromise let's just split the two into a novel and a backmutation file...

## 4 April 2019

Outputting the novel forward singletons to a TSV for talia

```{r echo=FALSE}

biallele.novel.counts.singleton.tbl <- biallele.novel.counts.singleton %>% filter(group=="mutant") %>% select(c("site_id", "chrom", "pos", `1_allele`, `2_allele`, "delta_bp")) %>% mutate(vartype=case_when(delta_bp < 0 ~ "del", delta_bp > 0 ~ "ins" )) %>%  rename("ref" = `1_allele`, alt=`2_allele`)
write.table(biallele.novel.counts.singleton.tbl , file = "novel_singleton_mcm5_indels.txt",row.names=FALSE, na="",col.names=FALSE, sep="\t")

```
uh, why doesn't this write to the  working directory??? (also adds quot marks around all the non-numerics, gross)

```{bash eval=FALSE}
join -1 3 -2 2 <( LANG=en_EN sort -k 3 novel_singleton_mcm5_indels.txt ) <(grep -v "#" variants/mutant.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.vcf | LANG=en_EN sort -k 2 ) | tr " " "\t" | awk '{if($3==$8)print;}' | cut -f 1-7,16- |sort -k3,3 -k1n,1  > novel_singleton_mcm5_indels.augmented.txt
```

UGH I remember this sort/join incompatibility now..........

https://unix.stackexchange.com/questions/12942/join-file-2-not-in-sorted-order

Hmm, why does one individual have nearly half the novel singletons???

## 11 April 2019

Ok gonna look at chrX... It looks like the main changes will be in the Winnower and Novelist rules. 

Winnower: add --chr chrX
Novelist: as is for autosomes. For X, we still need to demand that the parents are homozygous, and exactly one offspring is homozygous (not het) for the other allele. This is b/c the offspring are all male. If there are heterozygous variants there is a problem??? 

Actually no, that currently happens in the .Rmd; Novelist is fine as-is. 

Might as well use the uncorrected number in some plots (eg section 3.1). When it comes time to scan for novel variants, going to reset the chrX mac to floor(mac/2). (if there is a mac of 1 on a chrX, it gets rounded down to zero. )

Actually I guess it would be ceil() when it comes to the back mutations. 


```{r eval=FALSE}

mutate( mac = case_when( chrom == "chrX" ~ floor(mac/2), TRUE ~ as.double(mac)))

```

Wait.... this isn't going to let me differentiate between one 1/1 and two 0/1 's :(

## 12 April 2019 

Going to take a look at the unseemly amount of heterozygosity on the chrX. Using Kevin Blighe's code: 
https://www.biostars.org/p/291147/

https://stackoverflow.com/questions/19408649/pipe-input-into-a-script

```{bash eval=FALSE}
cat scripts/madHetter.sh 
```

```
#!/bin/bash
awk -F"\t" '{line=$0} BEGIN {
        print "CHR\tPOS\tID\tREF\tALT\tAltHetCount\tAltHomCount\tRefHomCount"
    } !/^#/ {
        if (gsub(/,/, ",", $5)==0) {
            print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t" gsub(/0\|1|1\|0|0\/1|1\/0/,"") "\t" gsub(/1\/1|1\|1/,"") "\t" gsub(/0\/0|0\|0/,"")
        } else if (gsub(/,/, ",", $5)==1) {
            print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t" gsub(/1\/0|0\/1|1\|0|0\|1|1\|2|2\|1|1\/2|2\/1/,"")","gsub(/2\/0|0\/2|2\|0|0\|2|1\|2|2\|1|1\/2|2\/1/,"",line) "\t" gsub(/1\/1|1\|1/,"")","gsub(/2\/2|2\|2/,"") "\t" gsub(/0\/0|0\|0/,"")
        }
    }'
```


```{bash eval=FALSE}
cat variants/mutant.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.vcf | sh scripts/madHetter.sh | grep -v "AltHetCount" | nl -n rn | awk '{print $2,$3,$3+1,$1,"0",$4,$5,$6,$7,$8,$9}'  | tr " " "\t" >  analysis/mutant.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.het.bg

cat variants/control.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.vcf | sh scripts/madHetter.sh | grep -v "AltHetCount" | nl -n rn | awk '{print $2,$3,$3+1,$1,"0",$4,$5,$6,$7,$8,$9}'  | tr " " "\t" >  analysis/control.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.het.bg

```

```{r echo=FALSE}
library("ggbio")
library("rtracklayer")

cntrl.hetZyggy.bg <- import.bedGraph("analysis/control.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.het.bg")

```



May be worth looking at the freebayes --cnv-map argument. 

## 15 April 2019



```{r echo=FALSE}

names(mcols(cntrl.hetZyggy.bg)) <- c("score", "zero", "na", "ref", "alt", "het", "homAlt", "homRef")
mcols(cntrl.hetZyggy.bg) <- mcols(cntrl.hetZyggy.bg) %>% as_tibble()  %>%  select(-c("zero", "na"))
cntrl.hetZyggy.bg$treat <- as.factor("control")

mut.hetZyggy.bg <- import.bedGraph("analysis/mutant.vs_dm6.bwaUniq.alleleCounts.simpleIndels.dpthFilt.biallelic.universal.het.bg")
names(mcols(mut.hetZyggy.bg)) <- c("score", "zero", "na", "ref", "alt", "het", "homAlt", "homRef")
mcols(mut.hetZyggy.bg) <- mcols(mut.hetZyggy.bg) %>% as_tibble()  %>%  select(-c("zero", "na"))
mut.hetZyggy.bg$treat <- as.factor("mutant")

hetZyggy.bg <- c(cntrl.hetZyggy.bg,mut.hetZyggy.bg)
hetZyggy.bg$tot <- hetZyggy.bg$het + hetZyggy.bg$homAlt + hetZyggy.bg$homRef 
hetZyggy.bg$zyggy <- (hetZyggy.bg$het)/hetZyggy.bg$tot
hetZyggy.bg$chrom <- hetZyggy.bg@seqnames

```

```{r echo=FALSE}

ggplot(mcols(hetZyggy.bg) %>% as.data.frame()) + geom_violin(aes(x=chrom, y=zyggy, fill=treat))

```

```{r echo=FALSE}

#autoplot(hetZyggy.bg, aes(y=zyggy, color=treat), geom='line') + facet_wrap(~seqnames, scales = "free_x") + labs(y="", title = "") + theme_clear() 

```
I'd need to window the bedgraph actually but the violin plot looks not tooo bad. 

Need to implement the --cnv-map





Corbin wants me to look and SNPs too





